State of babel-eslint blog post (#2291)
babel/website: website/blog/2020-07-13-the-state-of-babel-eslint.md

categories: announcements
share_text: "The State of babel-eslint"
---

## The Past

Existing as a compatibility layer between Babel and ESLint – two projects maintained by two separate teams – `babel-eslint` has been a difficult package to maintain since the beginning. Some of the challenges the team has faced while maintaining `babel-eslint`:

* `babel-eslint` has historically been maintained by the Babel team, and ensuring that ESLint's core rules (which is released every two weeks) work with experimental syntax is a monumental task.
* Babel's internal AST representation is different from [ESTree](https://github.com/estree/estree), the AST specification used by ESLint. As a result, Babel's parser has a plugin that outputs ASTs that are ESTree-compatible.
* While Babel enables the use of experimental syntax, ESLint core only supports syntax once it reaches [Stage 4](https://tc39.es/process-document/) and is finalized in the spec. This means that ESLint's core rules aren't designed to work with syntax that is Stage 3 and below. Additionally, ESLint has been blocked by the finalizing of the ESTree spec, leading to an even larger discrepancy in the syntax the two projects support.

> * The version of Babel's parser that `babel-eslint` relies on is a direct dependency in `package.json` and the plugins enabled are hardcoded in project's source code, leading to a potential mismatch in versions and enabled language features between compiling and linting. This has historically led to a lot of confusion and hard-to-debug issues, and we believe this is a big win for both maintainers and end users.

Reviewer: ```suggestion
* The version of Babel's parser that `babel-eslint` relies on is a direct dependency in its `package.json` and the plugins it enables are hardcoded, leading to a potential mismatch in versions and enabled language features between compiling and linting. This has historically led to a lot of confusion and hard-to-debug issues, and we believe this is a big win for both maintainers and end users.
```

* `@babel/eslint-parser` will require `@babel/core` as a peer dependency and will now use Babel core's APIs to read and apply your Babel configuration. This means that the same version of Babel with the same settings will be used for both compiling and linting. This is consistent with what we recommend and do with other packages in the Babel ecosystem.
* `@babel/eslint-parser` will live in the main [`babel/babel`](https://github.com/babel/babel/tree/main/eslint/babel-eslint-parser) monorepo with other Babel packages. We hope this will help to mitigate some of the syncing issues `babel-eslint` has seen in the past and allow us to lint the repo with the latest source code in GitHub, shortening the feedback loop of how changes in Babel affect linting.
* [ESTree](https://github.com/estree/estree) - the AST specification that ESLint follows - has recently become active again, and the Babel team is now actively involved in the project. This will hopefully allow us to share what we learn about the AST representation of experimental syntax and help make the integration between tools easier.

> * Kai Cataldo has been helping out with `babel-eslint` for a number of years now, and as a result of being a maintainer of both Babel and ESLint, is uniquely positioned to focus on the interoperability of these two tools. Over the past few months, he has been working on the code changes discussed above and will continue to support the new packages once they are released.

Reviewer: Should we link to your twitter or GitHub account?

## The Future

We believe that these packages are ready to be released. Because of all the integrations and projects involved (Babel, ESLint, Prettier, frameworks, various plugins, text editor integrations, command-line tools, etc.), we're sure there will be things that need to be fixed, and we plan to iterate quickly.

Please note that `@babel/eslint-parser` will rely on `@babel/core` as a peer dependency and this package must be in your project's `node_modules`.

Once you have ensured that `@babel/core` has been installed, you can run the following commands to upgrade from `babel-eslint` and `babel-eslint-plugin` to the new packages:

npm

```sh
npm uninstall babel-eslint babel-eslint-plugin
npm install --save-dev @babel/eslint-parser @babel/eslint-plugin
```

Yarn

```sh
yarn remove babel-eslint babel-eslint-plugin
yarn add --dev @babel/eslint-parser @babel/eslint-plugin
```

.eslintrc.js

```diff
module.exports = {
---  parser: "babel-eslint",
+++  parser: "@babel/eslint-parser"
    plugins: [
---   "babel"
+++   "@babel
    ]
};
```

> `@babel/eslint-parser` expects a standard Babel configuration file (`.babelrc` or `babel.config.js`). For cases where users might not want to use a Babel configuration file or are running Babel through another tool (such as Webpack with babel-loader), `@babel/eslint-parser` also supports applying configuration through your ESLint configuration. Please see [the readme](https://github.com/babel/babel/tree/main/eslint/babel-eslint-parser#additional-parser-configuration) for more details.

Reviewer: Should we link to docs for config file?

https://babeljs.io/docs/en/configuration
